File: 261-Graph-Valid-Tree/UnionFind.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n - 1:
            return False
        uf = UnionFind(n)
        for a, b in edges:
            if not uf.merge(a, b):
                return False
        return True


class UnionFind:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.size = [1] * n

    def root(self, i: int) -> int:
        if self.parent[i] != i:
            self.parent[i] = self.root(self.parent[i])
        return self.parent[i]

    def merge(self, i: int, j: int) -> bool:
        i, j = self.root(i), self.root(j)
        if i == j:
            return False
        if self.size[i] < self.size[j]:
            i, j = j, i
        self.parent[j] = i
        self.size[i] += self.size[j]
        return True

File: 261-Graph-Valid-Tree/UnionFind.kt
class Solution {
    fun validTree(n: Int, edges: Array<IntArray>): Boolean {
        if (edges.size != n - 1) {
            return false
        }
        
        val unionFind = UnionFind(n)
        for (edge in edges) {
            if (!unionFind.merge(edge[0], edge[1])) {
                return false
            }
        }
        
        return true
        
    }
}

class UnionFind(num: Int) {
    private val parent: IntArray
    private val size: IntArray
    
    init {
        parent = IntArray(num) {i -> i}
        size = IntArray(num) {1}
    }
    
    private fun root(node: Int): Int {
        if (node != parent[node]) {
            parent[node] = root(parent[node])
        }
        return parent[node]
    }
    
    fun merge(i: Int, j: Int): Boolean {
        var iRoot = root(i)
        var jRoot = root(j)
        
        if (iRoot == jRoot) {
            return false
        }
        
        if (size[iRoot] < size[jRoot]) {
            val tmp = jRoot
            jRoot = iRoot
            iRoot = tmp
        }
        
        parent[jRoot] = iRoot
        size[iRoot] += size[jRoot]
        
        return true
        
    }
    
}
